### ISSUE TITLE
Set alloc without initialize produces a broken Set

### ISSUE THREAD

**Issue by headius** [jruby/jruby#8360]
Our Set implementation has been a Java-based extension for some time. In order to remain compatible with CRuby, we allocate an internal Hash for it to provide to consumers that expect the `@hash` variable to be present.

This Hash is normally created during initialization of a new Set object, but if initialization is omitted by allocating a Set directly (`Set.allocate`) it produces a Set that has a null Hash and most accesses will fail with a NullPointerException:

```
$ jruby -rset -e 'Set.allocate << 1'
Unhandled Java exception: java.lang.NullPointerException: Cannot invoke "org.jruby.RubyHash.fastASetCheckString(org.jruby.Ruby, org.jruby.runtime.builtin.IRubyObject, org.jruby.runtime.builtin.IRubyObject)" because "this.hash" is null
java.lang.NullPointerException: Cannot invoke "org.jruby.RubyHash.fastASetCheckString(org.jruby.Ruby, org.jruby.runtime.builtin.IRubyObject, org.jruby.runtime.builtin.IRubyObject)" because "this.hash" is null
                       addImpl at org/jruby/ext/set/RubySet.java:635
                           add at org/jruby/ext/set/RubySet.java:630
                          call at org/jruby/ext/set/RubySet$INVOKER$i$1$0$add.gen:-1
                          call at org/jruby/internal/runtime/methods/AliasMethod.java:106
                  cacheAndCall at org/jruby/runtime/callsite/CachingCallSite.java:466
                          call at org/jruby/runtime/callsite/CachingCallSite.java:244
                          call at org/jruby/runtime/callsite/ShiftLeftCallSite.java:32
                          call at org/jruby/runtime/callsite/CachingCallSite.java:61
                          call at org/jruby/runtime/callsite/ShiftLeftCallSite.java:21
  invokeOtherOneFixnum1:\=\^\^ at -e:1
                   RUBY$script at -e:1
```

Forcing the Hash to be constructed from the Set constructor works around this, but may end up double-allocating the Hash since it will be allocated later for `initialize`:

```diff
diff --git a/core/src/main/java/org/jruby/ext/set/RubySet.java b/core/src/main/java/org/jruby/ext/set/RubySet.java
index b49e344175..832c52a94e 100644
--- a/core/src/main/java/org/jruby/ext/set/RubySet.java
+++ b/core/src/main/java/org/jruby/ext/set/RubySet.java
@@ -102,6 +102,8 @@ public class RubySet extends RubyObject implements Set {
 
     protected RubySet(Ruby runtime, RubyClass klass) {
         super(runtime, klass);
+
+        allocHash(runtime);
     }
 
     /*
```

This causes a second breakage in the example from #8352 because Psych uses `allocate` to construct new object instances, inserts them into an internal hash, and then proceeds to initialize them from the YAML stream.

There is no obvious workaround for this currently.

**Comment by headius** [comment]
@kares Perhaps we should just move the creation of the hash into the constructor and make it final? I can figure this out myself but since you wrote this Java-based Set maybe you know how to fix this quickly?

**Comment by kares** [comment]
I tried and can get the `allocate` working in a way where the hash is pre-allocated and avoid the 

but I do not think `allocate` is the issue, with YAML. MRI fails bad on trying to do anything with a `Set.allocate`

`ruby -rset -e 'Set.allocate << 1'`
```
/opt/local/rvm/rubies/ruby-3.1.4/lib/ruby/3.1.0/set.rb:522:in `add': undefined method `[]=' for nil:NilClass (NoMethodError)

    @hash[o] = true
         ^^^^^
	from -e:1:in `<main>'
```

as far as I understand we have YAML working for Set, it's only cyclic references to a set that are the problem...

**Comment by kares** [comment]
seems like we need a new [snakeyaml-engine for this](https://bitbucket.org/snakeyaml/snakeyaml-engine/pull-requests/16/diff), which isn't part of any release, yet. 
not going into building that myself, can try again once it's released...

**Comment by headius** [comment]
@kares So something is different with our set when used in a cyclic reference in Psych then... fine to wait until we have a patch from @asomov for the full SnakeYAML-Engine issue and then we'll figure out what else needs to chance in Psych.

**Comment by headius** [comment]
@kares Is there anything to do for this or is it an invalid bug? I think you have shown the real problem is nested Set in YAML, not specifically a problem with allocate-without-initialize as described.

We are still waiting on the updated SnakeYAML-Engine release.

**Comment by headius** [comment]
I'm going to close this in favor of #8393 which is the only thing we should fix right now. set.rb clearly also produces a broken Set when it is allocated but not initialized, and the fix in #8394 will emulate its behavior for Set#hash (it just hashes the nil `@hash` value).
